### PR TITLE
[MIR] Fix equality checks in matching code

### PR DESCRIPTION
**Pull request by nagisa**
This is not a fix to checks themselves per se (though we still use `Eq` MIR test instead of calling `PartialEq::eq`), but rather how we handle items we encounter in pattern position.

Previously we would just call `PartialEq` with the constant and the matchee, but now we essentially inline the constant instead. E.g. these two snippets are functionally equivalent at MIR level:

```
match val { Some(42) => true, _ => false }
```

and

```
const SECRET: Option<u8> = Some(42);
match val { SECRET => true, _ => false }
```

This approach also allows for more optimizations of matches. I.e. It can now exploit `SwitchInt` to switch on number inside a `Some` regardless of whether the value being an item or not.

This is based on @tsion’s already approved PR so I could reuse the file for more tests.

r? @eddyb 
cc @nikomatsakis @tsion

### CODE REVIEW

File: src/librustc_mir/hair/cx/pattern.rs [hunk 1]
```diff
@@ -88,22 +88,9 @@ impl<'patcx, 'cx, 'tcx> PatCx<'patcx, 'cx, 'tcx> {
...
-                                    let substs = self.cx.tcx.mk_substs(Substs::empty());
-                                    Literal::Item {
-                                        def_id: def_id,
-                                        kind: ItemKind::Constant,
-                                        substs: substs
-                                    }
-                                };
-                                PatternKind::Constant { value: literal }
+                                let pat = const_eval::const_expr_to_pat(self.cx.tcx, const_expr,
+                                                                        pat.span);
```
Comment (eddyb) [inline src/librustc_mir/hair/cx/pattern.rs]: Would be nice if we can use HAIR to avoid allocating here.

Comment (eddyb) [inline src/librustc_mir/hair/cx/pattern.rs]: Nevermind, that would overcomplicate too many things.

**Comment by eddyb** [comment]
Given that this is how we implement matching on constants currently, I don't see any problems with this.
@bors r+

**Comment by nikomatsakis** [comment]
On Sun, Jan 03, 2016 at 01:21:16PM -0800, Eduard-Mihai Burtescu wrote:
> Given that this is how we implement matching on constants currently, I don't see any problems with this.

Agreed (though I didn't yet look at the patch). I'd prefer to change
how we implement constants, but it's a separate thing.

**Pull request merged**